### PR TITLE
Multi-cursor: queue implementation

### PR DESCRIPTION
**Pull request by yycptt**
**What changed?**
- Multi-cursor: queue implementation

**Why?**
- Multi-cursor impl

**How did you test it?**
- Unit test

**Potential risks**
- Code not wired up yet.

**Is hotfix candidate?**
- No

### CODE REVIEW

File: service/history/queues/queue_base.go [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+	))
+}
+
+func (p *queueBase) processNewRange() {
+	newMaxKey := p.shard.GetQueueExclusiveHighReadWatermark(
+		p.category,
+		p.shard.GetClusterMetadata().GetCurrentClusterName(),
+	)
+
+	if !p.nonReadableScope.CanSplitByRange(newMaxKey) {
```
Comment (yux0) [inline service/history/queues/queue_base.go:261]: is the non readable scope the same as the write scope?

File: service/history/queues/queue_base.go [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+
+	// NOTE: Must range complete task first.
+	// Otherwise, if state is updated first, later deletion fails and shard get reloaded
+	// some tasks will never be deleted.
+	if newInclusiveLowWatermark != tasks.MaximumKey && newInclusiveLowWatermark.CompareTo(p.inclusiveLowWatermark) > 0 {
+		persistenceMinTaskKey := p.inclusiveLowWatermark
+		persistenceMaxTaskKey := newInclusiveLowWatermark
+		if p.category.Type() == tasks.CategoryTypeScheduled {
+			persistenceMinTaskKey = tasks.NewKey(p.inclusiveLowWatermark.FireTime, 0)
+			persistenceMaxTaskKey = tasks.NewKey(newInclusiveLowWatermark.FireTime, 0)
```
Comment (yux0) [inline service/history/queues/queue_base.go:320]: nit: possible to have a key.Copy()?

File: service/history/queues/queue_immediate.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+		notifyCh: make(chan struct{}, 1),
+	}
+}
+
+func (p *immediateQueue) Start() {
+	if !atomic.CompareAndSwapInt32(&p.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	p.logger.Info("", tag.LifeCycleStarting)
```
Comment (yux0) [inline service/history/queues/queue_immediate.go:107]: add a component tag?

File: service/history/queues/queue_scheduled.go [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+func (p *scheduledQueue) lookAheadTask() {
+	lookAheadMinTime := p.nonReadableScope.Range.InclusiveMin.FireTime
+	lookAheadMaxTime := lookAheadMinTime.Add(p.options.MaxPollInterval())
+
+	ctx, cancel := newQueueIOContext()
+	defer cancel()
+
+	request := &persistence.GetHistoryTasksRequest{
+		ShardID:             p.shard.GetShardID(),
+		TaskCategory:        tasks.CategoryTimer,
```
Comment (yux0) [inline service/history/queues/queue_scheduled.go:228]: p.queueBase.Category()?

File: service/history/queues/queue_scheduled.go [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+		p.timerGate.Update(lookAheadMinTime)
+		p.logger.Error("Failed to load look ahead task", tag.Error(err))
+		return
+	}
+
+	if len(response.Tasks) == 1 {
+		p.timerGate.Update(response.Tasks[0].GetKey().FireTime)
+	}
+
+	// no look ahead task, wait for max poll interval or new task notification
```
Comment (yux0) [inline service/history/queues/queue_scheduled.go:246]: do you need to reset the timer gate?

File: service/history/queues/queue_base.go (hunk 2 above)
Comment (yycptt) [reply service/history/queues/queue_base.go:320]: Looks like it's not necessary here, just want to reset the value for taskID.

File: service/history/queues/queue_scheduled.go (hunk 5 above)
Comment (yycptt) [reply service/history/queues/queue_scheduled.go:251]: Good catch! Performed a lookahead in after processing poll timer as well.

File: service/history/queues/queue_immediate.go (hunk 3 above)
Comment (yycptt) [reply service/history/queues/queue_immediate.go:107]: Component tag will be added by the caller of NewImmdicate/ScheduleQueue.

File: service/history/queues/queue_base.go (hunk 1 above)
Comment (yycptt) [reply service/history/queues/queue_base.go:261]: Yes, logically they are the same. But the non-readable scope here will be sightly behind the one in shard info, as the boundary is updated later.

File: service/history/queues/queue_scheduled.go (hunk 4 above)
Comment (yycptt) [reply service/history/queues/queue_scheduled.go]: Updated to `p.category`

File: service/history/queues/queue_scheduled.go (hunk 5 above)
Comment (yycptt) [reply service/history/queues/queue_scheduled.go:251]: Will add a test for this case.

**Review by yux0 — APPROVED** [review]

**Pull request merged**